src/thriveopedia/fossilisation/FossilisationDialog.cs

```diff
@@ -0,0 +1,232 @@
+    }
+
+    private void OnNameTextEntered(string newText)
+    {
+        PerformValidation(newText);
+    }
+
+    private void PerformValidation(string text)
+    {
+        // Only defocus if the name is valid to indicate invalid namings to the player
+        if (Regex.IsMatch(text, Constants.SPECIES_NAME_REGEX))
+        {
+            speciesNameEdit.ReleaseFocus();
+        }
+        else
+        {
+            // Prevents user from doing other actions with an invalid name
+            GetTree().SetInputAsHandled();
+
+            // TODO: Make the popup appear at the top of the line edit instead of at the last mouse position
+            ToolTipManager.Instance.ShowPopup(TranslationServer.Translate("INVALID_SPECIES_NAME_POPUP"), 2.5f);
+
+            speciesNameEdit.GetNode<AnimationPlayer>("AnimationPlayer").Play("invalidSpeciesNameFlash");
+        }
+    }
+
+    private void OnRandomizeNamePressed()
+    {
+        GUICommon.Instance.PlayButtonPressSound();
+
+        var nameGenerator = SimulationParameters.Instance.NameGenerator;
+        var randomizedName = nameGenerator.GenerateNameSection() + " " +
+            nameGenerator.GenerateNameSection(null, true);
+
+        speciesNameEdit.Text = randomizedName;
+        OnNameTextChanged(randomizedName);
+    }
+
+    private void UpdateSpeciesPreview()
+    {
+        speciesPreview.PreviewSpecies = SelectedSpecies;
+
+        switch (SelectedSpecies)
+        {
+            case MicrobeSpecies microbeSpecies:
+            {
+                hexesPreview.PreviewSpecies = microbeSpecies;
+                break;
+            }
+        }
+    }
+
+    private void UpdateSpeciesDetails()
+    {
+        speciesDetailsLabel.ExtendedBbcode = TranslationServer.Translate("SPECIES_DETAIL_TEXT").FormatSafe(
+            SelectedSpecies.FormattedNameBbCode,
+            SelectedSpecies.ID,
+            SelectedSpecies.Generation,
+            SelectedSpecies.Population,
+            SelectedSpecies.Colour.ToHtml(),
+            string.Join("\n  ", SelectedSpecies.Behaviour.Select(b => b.Key + ": " + b.Value)));
+
+        switch (SelectedSpecies)
+        {
+            case MicrobeSpecies microbeSpecies:
+            {
+                speciesDetailsLabel.ExtendedBbcode += "\n" +
+                    TranslationServer.Translate("MICROBE_SPECIES_DETAIL_TEXT").FormatSafe(
+                        microbeSpecies.MembraneType.Name, microbeSpecies.MembraneRigidity,
+                        microbeSpecies.BaseSpeed, microbeSpecies.BaseRotationSpeed, microbeSpecies.BaseHexSize);
+                break;
+            }
+        }
+    }
+
+    private void OnCancelPressed()
+    {
+        Hide();
+    }
+
+    private void OnFossilisePressed()
+    {
+        GD.Print("Saving species " + SelectedSpecies.FormattedName);
+
+        if (FossilisedSpecies.CreateListOfSaves()
+            .Any(s => s == SelectedSpecies.FormattedName + Constants.FOSSIL_EXTENSION_WITH_DOT))
```

Comment:
I'm seeing this logic for the second time so I'd say that FossilizedSpecies should have a method to check by *name* if something is fossilized already (and it could take an optional fossils list if called in a loop for example for more efficiency).